### PR TITLE
[FLINK-6288] fix FlinkKafkaProducer's custom Partitioner is always invoked with number of partitions of default topic

### PR DESCRIPTION
**Pull request by fanyon**
1. add extra api addTopicPartitioner, user can use it to add special topic and partitioner
2. add topicPartitionerMap in FlinkKafkaProducerBase to store the topic and partitioner
3. add PartitionerInfo to manage the topic and partitioner info

### CODE REVIEW

**Comment by tzulitai** [comment]
Thanks for the PR @fanyon. I'll try to look at the changes soon.

**Comment by gyfora** [comment]
Hi,
Thanks for the PR!

The first problem I noticed with this approach is that it will not work if users want to partition dynamically created topics (my use case actually). 

We should have a default partitioner that could be applied to the unmatched topics and would always pass the correct partition number.

**Comment by fanyon** [comment]
@gyfora Thanks for your comment. And right, your question is very good. I originally thought that the user must be sure to know all the output topic when the job is submitted, but in the real business scenario, the data may be output to the dynamically generated topic. 

For the requirementof generate dynamic topic, I propose to adjust the open and partition api of KafkaPartitioner as follows:
1. The open method, remove the parameter int[] partitions, and will be opend once for each partitioner
public void open(int parallelInstanceId, int parallelInstances)
2. The partition method, add int[] partitions and target topic parameters
public int partition(T next, byte[] serializedKey, byte[] serializedValue, String topic, int[] partitions)

@gyfora @tzulitai What do you think of this? Please feel free to give any suggestions, thanks!

**Comment by gyfora** [comment]
How would this new API map to the current one in terms of backwards compatibility?

**Comment by fanyon** [comment]
The new API should in a new class, such as FlinkKafkaPartitioner. For the older KafkaPartitioner implementation, it will be delegated by FlinkKafkaAdpterPartitioner extends FlinkKafkaPartitioner, in which contains defaultTopicId/partitions and KafkaPartitioner delegate, and map the new api to the current one in terms. Of course, as it is now, default topic's partitions will be used for the diffrent target topics in FlinkKafkaAdpterPartitioner.

**Comment by gyfora** [comment]
I think this is reasonable as the current implementation doesnt work for dynamic new topics. (we should also deprecate the current one)

But let's hear what others say :)

**Comment by fanyon** [comment]
@gyfora For the method `int partition(T next, byte[] serializedKey, byte[] serializedValue, int partitionNum)` in KafkaPartitioner, the correct partition num of target topic can be used. But the KafkaPartitioner's partition id array has been initialized in `void open(int parallelInstanceId, int parallelInstances, int[] partitions)`, which will be executed once, so yes, the problem for dynamic new topics when user uses older KafkaPartitioner API in their older job will still exist, and I find it hard to solve this problem completely.

What do you think of this? @tzulitai

**Comment by tzulitai** [comment]
Nice following the discussions here :) Let me wrap up the discussion so far:

The old way -
```
interface KafkaPartitioner {
    void open(int[] partitions, int subtaskIndex, int numSubtasks);
    int partition(T record, byte[] key, byte[] value, int numPartitions);
}
```

The (last) proposed new way -
```
interface FlinkKafkaPartitioner {
    void open(int subtaskIndex, int numSubtask);
    int partition(T record, byte[] key, byte[] value, String targetTopic, int[] partitions)
}
```
and have an internal cache of partitioner informations: `Map<String, PartitionerInfo>`.
The `PartitionerInfo` can actually just be the partition id array, I don't think we need another wrapper class if we just need a single `FlinkKafkaPartitioner` per subtask for all (including dynamic) topics.

I like the proposal of the new partitioner, as then users do not need to provide multiple partitioners. Just the question with how well this works for the general use case, because then implementations of the new `partition` method need to handle different topics (which probably makes sense because we want to generally treat topics as dynamic anyways ..). The new way can also allow us to handle upscaled target topics in the future.

For migration, for the dummy wrapper delegation, I think we should just mimc the wrong, old behaviour. That was the behaviour it had always been anyways, so we should not try to alter the behaviour if the user is still using the old API. Deprecation and Javadoc message is responsible of pushing them to change to the new API.

**Comment by tzulitai** [comment]
One thing to be careful with, though:
Since now we're querying Kafka for partition metadata within the `invoke` method, the query must be handled robustly and make sure it doesn't result in unexpectedly longer checkpoint times by blocking the whole stream at the Kafka sink.

Most notably, we need to consider the corner case where Kafka isn't cooperating nicely:
1. how to handle arbitrary long response time for fetching the partition metadata?
2. how to handle the case where, due to some Kafka brokers temporary unavailable, the returned partitions is not complete?

For 2., I can also forsee that we have a separate "partitions update thread" that refreshes the `Map<String, int[]>` cache continuously at a fixed interval. This can also involve to a `FlinkKafkaPartitioner` that can provide dynamically changing `int[] partitions` when invoking the `partition` method.

Perhaps we shouldn't include that with this PR, as its orthogonal to the API change. Just some food for though :)

**Comment by fanyon** [comment]
@tzulitai Thank you for your reply. For 2, I think a new issue may be created later.

For 1, it is really a problem for which will block the running job.  There may be two ways:

1. Depend on the timeout mechanism of kafka. When fetching partition meta from kafka, some timeout configurations should be setted.
2. Using Future to get the partiton meta of kafka, and user can set the timeout with configuration.

For the 1th way, problem may still exist for network and other reasons, so I'm apt to use the 2ed way. 
cc @gyfora

**Comment by gyfora** [comment]
I liked the proposed API and I agree that it's probably best to keep the old behaviour for the deprecated API.

I don't think the Kafka partition info fetching should be a huge problem as it shouldnt happen too often and Kafka should be able to return the info if you can write to it. We of course need some timeout/retry mechanism to not fail unnecessarily.

The producer itself is not very resilient in case of errors in the current state as it can't really handle the async errors it will just shut throw them and fail.

**Comment by fanyon** [comment]
As discussed above, I have updated the API and code.  Please review the changes when you are free, and be free to give any suggestions to me, thanks! @gyfora @tzulitai

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java [hunk 1]
```diff
@@ -281,6 +307,47 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
 		}
 	}
 
+	protected int[] getPartitionsByTopic(String topic, KafkaProducer<byte[], byte[]> producer) {
+		Future<int[]> future = executor.submit(new PartitionMetaTask(topic, producer));
+
+		try {
+			return future.get(kafkaMetaTimeoutMs, TimeUnit.MILLISECONDS);
+		} catch (Exception e) {
+			throw new RuntimeException(e);
```
Comment (gyfora) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java]: Should we maybe retry here a few times?

Comment (fanyon) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java]: Yes, retry here will be nicer, I'll fix it, thanks :)

**Comment by gyfora** [comment]
@tzulitai should we try to get this in the release?

**Comment by tzulitai** [comment]
@fanyon I'll finally have some time to get back to this PR this week (perhaps over the next 2 days). Thanks a lot for your patience ...

@gyfora I'm personally a +1 to try to get this in the release because it really is a self-contained thing, but I think it'll probably depend on the status of the 1.3 release in the end.

**Comment by tzulitai** [comment]
@zjureel could you rebase the PR on the latest master? Otherwise I cannot review the PR like this.

**Comment by tzulitai** [comment]
seems like the relevant commits are e6ec702 and 64af26e.
Let me try to resolve this .. :)

**Comment by zjureel** [comment]
@tzulitai Thank you for your replay. 

For 1, the `ExecutorService` is used to control timeout of fetching kafka partitions. When fetch kafka partitions, a `Future` will be created and executed in `ExecutorService`, which will wait for some mills and throw exception for timeout.
For 2, I had deprecated constructors of 08 / 09/ 010 whose parameter is `KafkaPartitioner` and add the same constructors with parameter `FlinkKafkaPartitioner`.

I find the codes in master of apache/flink were modified relative large some days ago, and I try rebase these modification soon. I think you can review these issues after that, thank you.

**Comment by tzulitai** [comment]
@zjureel thanks. You would need a proper rebase: `git rebase master` when you finish your feature branch, instead of merging the latest master.

Regarding timeout: doesn't the Kafka client have built-in timeout functionality?

**Comment by tzulitai** [comment]
@zjureel the rebases don't seem to be done correctly. The PR should contain the diff commits only.
I'm not sure what went wrong, but perhaps the most easiest way right now is cherry-pick your diff commits on a new branch checkedout from the latest master.

**Comment by zjureel** [comment]
@tzulitai Thank you for your suggestion, and I think you are right. I will create a new PR from the master and  cherry-pick my commits for this issue soon.

**Comment by fanyon** [comment]
@tzulitai I have created a new PR [https://github.com/apache/flink/pull/3901](url), and I will close this PR, you can review the code there, thanks.

**Pull request closed without merging**